### ISSUE TITLE
Upload image always run out of disk

### ISSUE THREAD

**Issue by csharpknife2002** [kubevirt/containerized-data-importer#1172]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:
I have setup an OpenShift cluster on GCP with 3 master and 3 nested virtualization enabled work nodes. I also have a default storage class which use the gce-pd and kubevirt has been installed.

I try to use the virtctl image-upload command to upload two images, the virtual image is are 50Gi and 100Gi.

virtctl image-upload -n MyNameSpace \
--image-path=images/my-image.qcow2 \
--pvc-name=mypvc \
--access-mode=ReadWriteOnce \
--pvc-size= a size \
--uploadproxy-url=https://myCDIUrl \
--insecure \
--wait-secs=3600

The 50Gi image can be upload when I set the pvc-size to 60Gi or 80 Gi, but when I inspect the log I saw the same issue mentioned by this issue: https://github.com/kubevirt/containerized-data-importer/issues/1078

But I always fail to upload the 100Gi image, it stopped after 10 tries of 
`
E0411 02:19:13.748907       1 prlimit.go:164] qemu-img failed output is:
E0411 02:19:13.748941       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
`
When I set it 100Gi, it will complain the actual size is smaller than 100Gi and cannot shrink, which is understandable.

But when I try to increase the size to 103, 105, 120, 200, 250 and 300Gi, I kept getting the same error and the image cannot be upload successfully, always failed after ten try. I post one of the logs below:
`
I0410 22:54:02.565725       1 uploadserver.go:62] Upload destination: /data/disk.img
I0410 22:54:02.566034       1 uploadserver.go:64] Running server on 0.0.0.0:8443
I0410 22:54:09.464332       1 uploadserver.go:263] Content type header is ""
I0410 22:54:09.464366       1 data-processor.go:253] Calculating available size
I0410 22:54:09.467300       1 data-processor.go:261] Checking out file system volume size.
I0410 22:54:09.467349       1 data-processor.go:265] Request image size not empty.
I0410 22:54:09.467364       1 data-processor.go:270] Target size 126210318336.
I0410 22:54:09.467425       1 util.go:37] deleting file: /scratch/lost+found
I0410 22:54:09.469430       1 util.go:37] deleting file: /data/lost+found
I0410 22:54:09.492838       1 data-processor.go:183] New phase: TransferScratch
I0410 22:54:09.496218       1 util.go:169] Writing data...
I0410 23:11:02.146150       1 data-processor.go:183] New phase: Process
I0410 23:11:02.147812       1 data-processor.go:183] New phase: Convert
I0410 23:11:02.147824       1 data-processor.go:189] Validating image
I0411 00:26:48.885450       1 data-processor.go:183] New phase: Resize
W0411 00:26:48.995598       1 data-processor.go:240] Available space less than requested size, resizing image to available space 126210318336.
I0411 00:26:49.006675       1 data-processor.go:246] Expanding image size to: 126210318336
E0411 00:26:49.494610       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:49.495250       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:49.692427       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:49.692458       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:49.852632       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:49.852661       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.061156       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.061185       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.438958       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.438990       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.604134       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.604162       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.725479       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.725507       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.828539       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.828569       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:50.978019       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:50.978045       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:51.069668       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:51.069692       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:51.241257       1 prlimit.go:164] qemu-img failed output is:
E0411 00:26:51.241284       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0411 00:26:51.241317       1 data-processor.go:180] exit status 1
qemu-img execution failed
kubevirt.io/containerized-data-importer/pkg/system.ExecWithLimits
	/go/src/kubevirt.io/containerized-data-importer/pkg/system/prlimit.go:166
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize
	/go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:139
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
	/go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
kubevirt.io/containerized-data-importer/pkg/uploadserver.newUploadStreamProcessor
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:292
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandler
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:265
net/http.HandlerFunc.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:1995
net/http.(*ServeMux).ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2375
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:207
net/http.serverHandler.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2774
net/http.(*conn).serve
	/usr/lib/golang/src/net/http/server.go:1878
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1337
Error resizing image /data/disk.img
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize
	/go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:141
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
	/go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
kubevirt.io/containerized-data-importer/pkg/uploadserver.newUploadStreamProcessor
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:292
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandler
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:265
net/http.HandlerFunc.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:1995
net/http.(*ServeMux).ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2375
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:207
net/http.serverHandler.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2774
net/http.(*conn).serve
	/usr/lib/golang/src/net/http/server.go:1878
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1337
Resize of image failed
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:219
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
kubevirt.io/containerized-data-importer/pkg/uploadserver.newUploadStreamProcessor
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:292
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandler
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:265
net/http.HandlerFunc.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:1995
net/http.(*ServeMux).ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2375
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:207
net/http.serverHandler.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2774
net/http.(*conn).serve
	/usr/lib/golang/src/net/http/server.go:1878
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1337
Unable to resize disk image to requested size
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
	/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:174
kubevirt.io/containerized-data-importer/pkg/uploadserver.newUploadStreamProcessor
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:292
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandler
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:265
net/http.HandlerFunc.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:1995
net/http.(*ServeMux).ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2375
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	/go/src/kubevirt.io/containerized-data-importer/pkg/uploadserver/uploadserver.go:207
net/http.serverHandler.ServeHTTP
	/usr/lib/golang/src/net/http/server.go:2774
net/http.(*conn).serve
	/usr/lib/golang/src/net/http/server.go:1878
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1337
I0411 00:26:51.242402       1 util.go:37] deleting file: /scratch/tmpimage
E0411 00:27:30.470475       1 uploadserver.go:271] Saving stream failed: Unable to resize disk image to requested size: Resize of image failed: Error resizing image /data/disk.img: qemu-img execution failed: exit status 1
`
**What you expected to happen**:
I expected the second image (100Gi) be uploaded successfully

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`):
operator.cdi.kubevirt.io/createVersion: v2.2.0-5

- Kubernetes version (use `kubectl version`):
oc v3.6.0-alpha.2+3c221d5
kubernetes v1.6.1+5115d708d7
kubernetes v1.16.2
- Cloud provider or hardware configuration:
openshift 4.3.9 on gcp, 
- Install tools: openshift
- Others:

**Comment by csharpknife2002** [comment]
The part I don't understand is:
the virtual image size is 100Gi (107374182400 )
the pvc size is 120 Gi (128849018880),

when I inspect the available size myself in the pod, 
I can see the available size is 126210318336 (123252264*1024), which is same as the target size/available size report in the logs from the cdi-upload-server.

Why the resize still not work?


sh-4.4# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0   512G  0 disk 
|-sda1   8:1        384M  0 part 
|-sda2   8:2        127M  0 part 
|-sda3   8:3          1M  0 part 
`-sda4   8:4    0 511.5G  0 part 
sdb      8:16   0   120G  0 disk 
sdc      8:32   0   120G  0 disk 
sdd      8:48   0   103G  0 disk 
sde      8:64   0   103G  0 disk 
sdf      8:80   0   120G  0 disk /data
sdg      8:96   0   120G  0 disk /scratch
sh-4.4# df /data
Filesystem     1K-blocks  Used Available Use% Mounted on
/dev/sdf       123330112 61464 123252264   1% /data

**Comment by csharpknife2002** [comment]
when i tried with a pvc with 102.5 Gi.
the logs said : 
I0411 10:09:33.770389       1 data-processor.go:189] Validating image

I  inspected the /data folder, I saw the disk.img is already there and i tried to use df /data I saw the available size is continuously decreasing.. What is happen behind the scene? Since the disk.img size is not changing, who consume those spaces?


`
sh-4.4# ls -l -a /data
total 104858028
drwxr-xr-x. 2 root root         4096 Apr 11 10:09 .
drwxr-xr-x. 1 root root           44 Apr 11 09:50 ..
-rw-r--r--. 1 root root 107374182400 Apr 11 10:27 disk.img
`
`
sh-4.4# df /data -a
Filesystem     1K-blocks      Used Available Use% Mounted on
/dev/sdc       105784056 104919568    848104 100% /data
sh-4.4# df /data -a
Filesystem     1K-blocks      Used Available Use% Mounted on
/dev/sdc       105784056 104919576    848096 100% /data
sh-4.4# df /data -a
Filesystem     1K-blocks      Used Available Use% Mounted on
/dev/sdc       105784056 104919580    848092 100% /data
`
